### DISCUSSION TITLE
Navier-Stokes Boundary conditions

### DISCUSSION THREAD

**Discussion by lacombef** [mfem/mfem#3382] [Q&A] [answered]
Hi all,

Me again, in the Navier-Stokes miniapps, I see that most of the time, the boundary conditions are applied using an arbitrary function(x,t), that return either the pressure or velocity. All the boundaries use the same function and the boundaries are identified based on their position "x". In complex geometries, it is complicated to find a condition to identify each boundary. Maybe it's a dumb question, but is it possible to identify the surfaces by their attribute number or any form of Identification number ?

Thank you !

Francis

**Comment by jandrej** [comment, marked as answer]
Yes this is of course possible. Boundaries on meshes can be tagged by boundary attributes during the meshing process. You can look at https://github.com/mfem/mfem/blob/master/miniapps/navier/navier_turbchan.cpp#L154-L157 where I specify attributes instead of checking the topological location.

**Reply by lacombef** [reply]
Maybe I didn't understand the code, but in this example, it seems like you use the same function, "vel_wall", for all the boundaries. If vel_wall could have the boundary attribute number in argument, that would be very simple to define the BC.

**Reply by lacombef** [reply]
Nevermind, I found it. You can simply use this

   // Inlet is attribute 1.
   attr = 0 ; attr[0] = 1;
   flowsolver.AddVelDirichletBC(vel_1, attr) ;
   // Walls is attribute 3.
   attr = 0 ; attr[2] = 1 ;
   flowsolver.AddVelDirichletBC(vel_3, attr) ;

**Comment by giraldeau** [comment]
Using gmsh, we would setup BC using physical groups. I tried to define a very simple rectangular surface mesh with four distinct sides, and it seems that the `mesh.bdr_attributes` is empty.

* test.geo
```gmsh
SetFactory("OpenCASCADE");
Mesh.CharacteristicLengthMax = 0.1;
Mesh.SaveAll = 1;

Rectangle(1) = {0, 0, 0, 1, 0.5, 0};
Physical Curve("left", 5) = {4};
Physical Curve("top", 6) = {3};
Physical Curve("right", 7) = {2};
Physical Curve("bot", 8) = {1};
Physical Surface("Surf") = {1};
```

* generate mesh

```bash
gmsh -2 -format msh2 test.geo
```

Just load the mesh and display `mesh.bdr_attributes.Max()`

```cpp
   Mesh mesh;
   mesh.LoadFromFile(mesh_file);
   std::cout << "bdr_attributes: " << mesh.bdr_attributes.Max() << std::endl;
/*
Gmsh reader: all element attributes were zero.
MFEM only supports positive element attributes.
Setting element attributes to 1.

Assertion failed: (size > 0) is false:
 --> Array is empty with size 0
 ... in function: T mfem::Array<T>::Max() const [with T = int]
 ... in file: /home/francisg/git/mfem/general/array.cpp:70

Aborted (core dumped)
*/
```

Somehow, the physical groups are lost... Idealy, we would be able to retrieve the boundary ID corresponding to a named physical names, but that doesn't seems to be implemented at the moment.

**Reply by jandrej** [reply]
I know that gmsh attributes work. I don't see anything wrong with your input file though.

> Idealy, we would be able to retrieve the boundary ID corresponding to a named physical names, but that doesn't seems to be implemented at the moment

Correct. It is trivial to implement and orthogonal to the mesh reader capability internal functionality.

Can any GMSH expert chime in here? @bslazarov maybe?

**Reply by etiennespieser** [reply]
Hello all,

Latest gmsh file format is not supported in MFEM, see https://mfem.org/mesh-formats/.
What seems missing in your example is the line
Mesh.MshFileVersion = 2.2;

I hope that this answer, even if late, will be useful.

Étienne
